feat: support nullable vector field data

marcelo-cjl opened the pull request:
Supported types: FloatVector, BinaryVector, Float16Vector, BFloat16Vector,
Int8Vector and SparseFloatVector.

C++ layer changes:
- Add FieldDataVectorImpl class with LogicalToPhysicalMapping for mapping storage
- LogicalToPhysicalMapping uses adaptive strategy: map when valid_ratio < 10%, otherwise uses vector for O(1) lookup
- Override FillFieldData() in FieldDataVectorImpl to handle nullable vectors
- Remove "vector not support null" restriction in Util.cpp

Go layer changes:
- Add LogicalToPhysicalMapping struct with validCount and l2pMap
- Add ValidData, Nullable, L2PMapping fields to all vector structs
- Update GetRow() to use L2PMapping.GetPhysicalOffset() for physical index
- Update AppendRow()/AppendValidDataRows() to build mapping incrementally
- Update GetMemorySize() to include ValidData, Nullable and L2PMapping

Storage strategy: nullable vectors use sparse storage where Data array only contains non-null vectors, ValidData bitmap tracks null positions, L2PMapping translates logical offset to physical offset.

marcelo-cjl commented:
relate #45994

marcelo-cjl commented:
/ci-rerun-e2e-default

foxspy commented on internal/core/src/common/FieldDataInterface.h:
```diff
@@ -538,18 +567,58 @@ class FieldDataImpl : public FieldDataBase {
+    using FieldDataImpl<Type, is_type_entire_row>::FieldDataImpl;
+
+    const void*
+    RawValue(ssize_t offset) const override {
```
This offset shouldn’t depend on whether it’s valid or not — this change alters the interface’s definition.

foxspy commented on internal/core/src/common/FieldDataInterface.h:
```diff
@@ -538,18 +567,58 @@ class FieldDataImpl : public FieldDataBase {
+        return sizeof(Type) * this->length_ * dim;
+    }
+
+    int64_t
```
Same as above

marcelo-cjl replied:
done

foxspy commented on internal/storage/insert_data.go:
```diff
@@ -493,29 +530,70 @@ func (dst *SparseFloatVectorFieldData) AppendAllRows(src *SparseFloatVectorField
+	return len(data.Data) * 8 / data.Dim
+}
+
+func (data *FloatVectorFieldData) RowNum() int {
```
GetRowSize should also be updated when nullable vector is supported

marcelo-cjl committed:
Add nullable support for all vector field types

Supported types: FloatVector, BinaryVector, Float16Vector, BFloat16Vector,
Int8Vector and SparseFloatVector.

C++ layer changes:
- Add FieldDataVectorImpl class with LogicalToPhysicalMapping for sparse storage
- LogicalToPhysicalMapping uses adaptive strategy: map when valid_ratio < 10%,
  otherwise uses vector for O(1) lookup
- Override FillFieldData() in FieldDataVectorImpl to handle nullable vectors
- Move FieldDataSparseVectorImpl from FieldDataInterface.h to FieldData.h
- Remove "vector not support null" restriction in Util.cpp

Go layer changes:
- Add LogicalToPhysicalMapping struct with validCount and l2pMap
- Add ValidData, Nullable, L2PMapping fields to all vector structs
- Update GetRow() to use L2PMapping.GetPhysicalOffset() for physical index
- Update AppendRow()/AppendValidDataRows() to build mapping incrementally
- Update GetMemorySize() to include ValidData, Nullable and L2PMapping

Storage strategy: nullable vectors use sparse storage where Data array only
contains non-null vectors, ValidData bitmap tracks null positions,
L2PMapping translates logical offset to physical offset.

Signed-off-by: marcelo-cjl <marcelo.chen@zilliz.com>

marcelo-cjl commented on internal/core/src/common/FieldData.cpp:
```diff
@@ -83,6 +83,53 @@ FieldDataImpl<Type, is_type_entire_row>::FillFieldData(
+    if (valid_data) {
+        for (ssize_t i = 0; i < total_element_count; ++i) {
+            if ((valid_data[i >> 3] >> (i & 0x07)) & 1) {
+                valid_count++;
```
done

marcelo-cjl commented on internal/storage/insert_data.go:
```diff
@@ -454,29 +467,53 @@ type GeometryFieldData struct {
+	ValidCount int
+	Dim        int
+	Nullable   bool
 }
```
done
